### PR TITLE
Pedestrian crossing should be a separate TripLeg_Use

### PR DESCRIPTION
**Pull request by abhi2039**
# Pedestrian crossing should be a separate TripLeg_Use 

What issue is this PR targeting?  Issue #2942 

Added a separate use for pedestrian crossings. Updated narrative dictionary (English only for now and default to old behavior for other languages) to have instructions like "Turn right on the pedestrian crossing"

## Tasklist

 -  Added tests
 - [ ] Add #fixes with the issue number that this PR addresses
 - [ ] Update the docs with any new request parameters or changes to behavior described
 - [ ] Update the [changelog](CHANGELOG.md)

## Requirements / Relations

 Link any requirements here. Other pull requests this PR is based on?

### CODE REVIEW

**Review by dnesbitt61 — DISMISSED** [review]
Looks good to me from a data building side. I'll let others comment on the narrative/guidance updates.

**Review by dgearhart — CHANGES_REQUESTED** [review]
@abhi2039 I will take a look

File: proto/trip.proto [hunk 1]
```diff
@@ -57,6 +57,7 @@ message TripLeg {
     kPathUse = 27;
     kPedestrianUse = 28;
     kBridlewayUse = 29;
+    kPedestrianCrossing = 32;
```
Comment (dgearhart) [inline proto/trip.proto:60]: Add `Use` suffix to keep consistent

File: src/odin/narrativebuilder.cc [hunk 2]
```diff
@@ -4071,9 +4071,18 @@ NarrativeBuilder::FormStreetNames(const Maneuver& maneuver,
...
     if ((maneuver.travel_mode() == TripLeg_TravelMode_kPedestrian) && maneuver.unnamed_walkway()) {
-      street_names_string = empty_street_name_labels->at(kWalkwayIndex);
+      auto num_empty_street_labels = empty_street_name_labels->size();
+
+      // Pedestrian crossing string is only available for english locale. Default to kWalkwayIndex
+      // for other languages
+      bool apply_pedestrian_crossing_label =
+          maneuver.pedestrian_crossing() && kPedestrianCrossingIndex < num_empty_street_labels;
+      auto dictionary_index =
+          apply_pedestrian_crossing_label ? kPedestrianCrossingIndex : kWalkwayIndex;
```
Comment (dgearhart) [inline src/odin/narrativebuilder.cc:4084]: Don't need to do this - just have English placeholder in other language files

File: locales/en-US.json [hunk 3]
```diff
@@ -44,7 +44,8 @@
       "empty_street_name_labels": [
         "the walkway",
         "the cycleway",
-        "the mountain bike trail"
+        "the mountain bike trail",
+        "the pedestrian crossing"
```
Comment (dgearhart) [inline locales/en-US.json:48]: Thinking the en-US (the default) phrase should be `"the crosswalk"` as per: https://en.wikipedia.org/wiki/Pedestrian_crossing

File: test/narrative_dictionary.cc [hunk 4]
```diff
@@ -14,7 +14,8 @@ namespace {
 
 // Expected strings
 const std::vector<std::string> kExpectedEmptyStreetNameLabels = {"the walkway", "the cycleway",
-                                                                 "the mountain bike trail"};
+                                                                 "the mountain bike trail",
+                                                                 "the pedestrian crossing"};
```
Comment (dgearhart) [inline test/narrative_dictionary.cc:18]: Thinking the en-US (the default) phrase should be "the crosswalk" as per: https://en.wikipedia.org/wiki/Pedestrian_crossing

**Comment by dgearhart** [comment]
@abhi2039 nice job! I left a few comments
Could you please add a gurka test demonstrating the crosswalk use?
[Valhalla Gurka Tests](https://github.com/valhalla/valhalla/blob/master/test/gurka/README.md)
Thank you for contributing

File: src/odin/narrativebuilder.cc (hunk 2 above)
Comment (abhi2039) [reply src/odin/narrativebuilder.cc:4084]: Thanks for reviewing. Yeah, it does not make sense anymore. I had added it before I realized that if I don't add the corresponding strings in other locales, the tests will fail. I will get rid of it.

**Comment by abhi2039** [comment]
> @abhi2039 nice job! I left a few comments
> Could you please add a gurka test demonstrating the crosswalk use?
> [Valhalla Gurka Tests](https://github.com/valhalla/valhalla/blob/master/test/gurka/README.md)
> Thank you for contributing

Thank you, @dgearhart! Addressed your comments.

**Comment by abhi2039** [comment]
> @abhi2039 nice job! I left a few comments
> Could you please add a gurka test demonstrating the crosswalk use?
> [Valhalla Gurka Tests](https://github.com/valhalla/valhalla/blob/master/test/gurka/README.md)
> Thank you for contributing

Thank you, @dgearhart ! Addressed your comments.

File: test/gurka/test_crosswalk.cc [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+  const auto layout = gurka::detail::map_to_coordinates(ascii_map, 50);
+  auto map = gurka::buildtiles(layout, ways, {}, {}, "test/data/crosswalk");
+  auto result = gurka::do_action(valhalla::Options::route, map, {"A", "B"}, "pedestrian");
+
+  EXPECT_EQ(result.directions().routes_size(), 1);
+  EXPECT_EQ(result.directions().routes(0).legs_size(), 1);
+  gurka::assert::raw::expect_path(result, {"AB"});
+
+  auto leg = result.trip().routes(0).legs(0);
+  EXPECT_EQ(leg.node(0).edge().use(), valhalla::TripLeg_Use::TripLeg_Use_kPedestrianCrossingUse);
```
Comment (dgearhart) [inline test/gurka/test_crosswalk.cc:22]: could we test a path we traverse a crosswalk in the middle - walk down street and then turn left into the crosswalk, etc
You can have test check the actual instructions

Comment (abhi2039) [reply test/gurka/test_crosswalk.cc:22]: Yeah, I wanted to do that but unfortunately couldn't figure out a way to make the ways unnamed in this test. The changes I made are for only unnamed pedestrian crossings. For named pedestrian crossings, nothing changes.

Comment (dgearhart) [reply test/gurka/test_crosswalk.cc:22]: should be able to use `{"name", ""}`

Comment (abhi2039) [reply test/gurka/test_crosswalk.cc:22]: Cool! I will give that a go.

Comment (abhi2039) [reply test/gurka/test_crosswalk.cc]: All set now!

**Comment by abhi2039** [comment]
@dgearhart Does this look good to go?

**Comment by dgearhart** [comment]
> @dgearhart Does this look good to go?

I think so - I will try to take final pass today

**Comment by dgearhart** [comment]
@abhi2039 sorry ran out time Friday - I grabbed your branch and will review and run some tests this afternoon

**Comment by dgearhart** [comment]
@abhi2039 I did a small data conversion with the first route in the [pedestrian_routes.txt](https://github.com/valhalla/valhalla/blob/master/test_requests/pedestrian_routes.txt#L1) test file
image

```
1: Walk west on Kingston Street.
2: Turn right onto the crosswalk. 
```
This looks odd to me - since the user might not take the crosswalk in this scenario

Did you [run](https://github.com/valhalla/valhalla/blob/master/run_route_scripts/README.md) the pedestrian test files with new data and diff the results? If so, what were you seeing?

**Comment by abhi2039** [comment]
> @abhi2039 I did a small data conversion with the first route in the [pedestrian_routes.txt](https://github.com/valhalla/valhalla/blob/master/test_requests/pedestrian_routes.txt#L1) test file
> image
> 
> ```
> 1: Walk west on Kingston Street.
> 2: Turn right onto the crosswalk. 
> ```
> 
> This looks odd to me - since the user might not take the crosswalk in this scenario
> 
> Did you [run](https://github.com/valhalla/valhalla/blob/master/run_route_scripts/README.md) the pedestrian test files with new data and diff the results? If so, what were you seeing?

I have not run the pedestrian test files with new data. I have tested it out on the internal fork of Valhalla (at work) which already has these changes. 
The right turn maneuver on San Jose Avenue is starting on the crosswalk and then moves to the sidewalk. Is the concern that the path is transitioning from the road to sidewalk and there is no need to call out crosswalk?

**Comment by dgearhart** [comment]
> The right turn maneuver on San Jose Avenue is starting on the crosswalk and then moves to the sidewalk. Is the concern that the path is transitioning from the road to sidewalk and there is no need to call out crosswalk?

Yes

**Comment by dgearhart** [comment]
> Yeah, I think that's more intuitive. In that case, the instruction will be "Turn right onto the walkway". Basically, I need to add some logic to detect if the maneuver is from a road to a sidewalk. In those cases, we don't need to call out crosswalks. When the maneuver is from a sidewalk to a crosswalk, then the instruction "Turn right onto the crosswalk" makes more sense.
> I think we are on the same page but let me know if I misunderstood something.

yes - exactly.
Also, what are your thoughts on how explicit the instructions should be for each crosswalk traversal? If a user was walking straight down a sidewalk and traversed 3 crosswalks at 3 different intersections - should each one be called out?

**Comment by abhi2039** [comment]
> > Yeah, I think that's more intuitive. In that case, the instruction will be "Turn right onto the walkway". Basically, I need to add some logic to detect if the maneuver is from a road to a sidewalk. In those cases, we don't need to call out crosswalks. When the maneuver is from a sidewalk to a crosswalk, then the instruction "Turn right onto the crosswalk" makes more sense.
> > I think we are on the same page but let me know if I misunderstood something.
> 
> yes - exactly.
> Also, what are your thoughts on how explicit the instructions should be for each crosswalk traversal? If a user was walking straight down a sidewalk and traversed 3 crosswalks at 3 different intersections - should each one be called out?

@dgearhart I updated the logic and added a couple of test points. Please take a look whenever you get a chance. Thanks again for your thorough review.

**Review by dgearhart — COMMENTED** [review]
@abhi2039 looks good - just need to get CI passing

**Comment by abhi2039** [comment]
> @abhi2039 looks good - just need to get CI passing

@dgearhart Thanks! Just to be sure - is there anything you need from me?

**Comment by kevinkreiser** [comment]
@abhi2039 looks like you just need to resolve the conflict with master. in a recent pr we changed `Maneuver` constructor in the same place you did so you'll just have to resolve that yet!

**Review by dgearhart — APPROVED** [review]
Thank you @abhi2039  :ship:

**Comment by grzezlo** [comment]
Thanks for introducing this great feature.
Unfortunately, I found some problem related to detecting the crosswalks.
If the crosswalk is the line, tagged as highway=footway and highway=crossing, everything works well, but often the pedestrian crossing is a part of longer footway.
Example:
at lat:52.321434, lon:20.971978 there is a node[1] in OpenStreetMap tagged highway=crossing.
Way[2] which crosses this node is tagged as highway=footway.
So the line 1312 of graph.lua can not detect this situation, because one of required tags is on the node, and other, on the way.
```
   elseif (kv["highway"] == "footway" and kv["footway"] == "crossing") then
     use = 32
```     
Can it be addressed somehow, or the limitations of processing in graph.lua makes it impossible?
[1]: https://www.openstreetmap.org/node/1459336494
[2]: https://www.openstreetmap.org/way/176939492

**Comment by abhi2039** [comment]
> Thanks for introducing this great feature.
> Unfortunately, I found some problem related to detecting the crosswalks.
> If the crosswalk is the line, tagged as highway=footway and highway=crossing, everything works well, but often the pedestrian crossing is a part of longer footway.
> Example:
> at lat:52.321434, lon:20.971978 there is a node[1] in OpenStreetMap tagged highway=crossing.
> Way[2] which crosses this node is tagged as highway=footway.
> So the line 1312 of graph.lua can not detect this situation, because one of required tags is on the node, and other, on the way.
> 
> ```
>    elseif (kv["highway"] == "footway" and kv["footway"] == "crossing") then
>      use = 32
> ```
> 
> Can it be addressed somehow, or the limitations of processing in graph.lua makes it impossible?
> [1]: https://www.openstreetmap.org/node/1459336494
> [2]: https://www.openstreetmap.org/way/176939492

Sorry for the late response. I cannot think of a way to do this. @dgearhart , @kevinkreiser any suggestions?

**Pull request merged**